core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -443,61 +444,40 @@ sealed abstract class Doc extends Product with Serializable {
 
-  /**
-   * Convert this Doc to a single-line representation.
-   *
-   * All newlines are replaced with spaces (and optional indentation
-   * is ignored). The resulting Doc will never render any newlines, no
-   * matter what width is used.
-   */
-  def flatten: Doc = flattenBoolean._1
-
-  // return the flattened doc, and if it is different
-  private def flattenBoolean: (Doc, Boolean) = {
-
-    type DB = (Doc, Boolean)
+  // Try to put the document on a single line.  Will succeed iff there are no
+  // hard line breaks.
+  // Invariant: If flatten(d) == Some(d') then d' has no FlatAlt nodes
+  private[paiges] def flatten: Option[Doc] = {
+    sealed trait Flatten extends Product with Serializable
+    final case class Flat(doc: Doc, changed: Boolean) extends Flatten {
+      def cat(that: Flat): Flat = Flat(Concat(doc, that.doc), changed || that.changed)
+    }
+    case object No extends Flatten
 
-    def finish(last: DB, front: List[DB]): DB =
-      front.foldLeft(last) {
-        case ((d1, c1), (d0, c2)) => (Concat(d0, d1), c1 || c2)
-      }
+    def finish(last: Flat, front: List[Flat]): Flatten = front.foldLeft(last) { case (b, a) => a cat b }
 
     @tailrec
-    def loop(h: DB, stack: List[DB], front: List[DB]): DB =
-      h._1 match {
-        case Empty | Text(_) =>
-          stack match {
-            case Nil => finish(h, front)
-            case x :: xs => loop(x, xs, h :: front)
-          }
-        case l@Line(_) =>
-          val next = l.asFlatDoc
-          val change = (next, true)
-          stack match {
-            case Nil => finish(change, front)
-            case x :: xs => loop(x, xs, change :: front)
-          }
-        case Nest(i, d) =>
-          loop((d, h._2), stack, front) // no Line, so Nest is irrelevant
-        case Align(d) =>
-          loop((d, h._2), stack, front) // no Line, so Align is irrelevant
-        case l@LazyDoc(_) => loop((l.evaluated, h._2), stack, front)
-        case Union(a, _) => loop((a, true), stack, front) // invariant: flatten(union(a, b)) == flatten(a)
-        case Concat(a, b) => loop((a, h._2), (b, h._2) :: stack, front)
+    def loop(h: Flat, stack: List[Flat], front: List[Flat]): Flatten =
+      h.doc match {
+        case Empty | Text(_) => stack match {
+          case Nil => finish(h, front)
+          case x :: xs => loop(x, xs, h :: front)
+        }
+        case Line => No
+        case Nest(_, d) => loop(Flat(d, h.changed), stack, front) // no Line, so Nest is irrelevant
+        case Align(d) => loop(Flat(d, h.changed), stack, front) // no Line, so Align is irrelevant
+        case Union(a, _) => loop(Flat(a, true), stack, front) // invariant: flatten(union(a, b)) == flatten(a)
+        case Concat(a, b) => loop(Flat(a, h.changed), Flat(b, h.changed) :: stack, front)
+        case FlatAlt(_, b) => loop(Flat(b, true), stack, front)
+        case LazyDoc(a) => loop(Flat(a(), false), stack, front)
```

Comment:
we should use `l@LazyDoc => l.evaluated` here so we cache the evaluation if it has already happened. We don't want to evaluate those more than once.